### DISCUSSION TITLE
Templates in the data: block of Dev->Services aren't casted/converted to the expected type

### DISCUSSION THREAD

**Discussion by rccoleman** [home-assistant/frontend#7497] [Other feature requests]
## The problem

Templates in the `data:` block of Dev->Services aren't casted/coerced into the expected field type, but they are in a script.  This example from the `input_datetime` docs works in a script, but not in Dev->Services:

```
- service: input_datetime.set_datetime
  data:
    entity_id: input_datetime.XXX
    timestamp: "{{ now().timestamp() }}"
```

In Dev->Services, it results in the following error:

`voluptuous.error.MultipleInvalid: expected float for dictionary value @ data['timestamp']`

Templates always return strings, but it looks like they're converted to the expected type in a script, but not in the developer tools.

## Environment

- Home Assistant Core release with the issue: 0.115.2
- Last working Home Assistant Core release (if known): Unknown
- Operating environment (OS/Container/Supervised/Core): OS
- Integration causing this issue: input_datetime
- Link to integration documentation on our website: https://www.home-assistant.io/integrations/input_datetime/

## Problem-relevant `configuration.yaml`

I created an `input_datetime` (helper) in the UI for testing, and used this script:

```yaml
  test:
    sequence:
      - service: input_datetime.set_datetime
        data:
          entity_id: input_datetime.date_test
          timestamp: "{{ now().timestamp() }}"
```

That works, while putting the following in the data section of Dev->Services with `input_datetime.set_datetime` results in the error mentioned:

```
entity_id: input_datetime.date_test
timestamp: "{{ now().timestamp() }}"
```

## Traceback/Error logs

```txt
2020-09-20 08:59:17 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.140399136576464] expected float for dictionary value @ data['timestamp']
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 137, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1269, in async_call
    processed_data = handler.schema(service_data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 208, in __call__
    return self._exec((Schema(val) for val in self.validators), v)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 287, in _exec
    raise e if self.msg is None else AllInvalid(self.msg, path=path)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/validators.py", line 283, in _exec
    v = func(v)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 817, in validate_callable
    return schema(data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 272, in __call__
    return self._compiled([], data)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 594, in validate_dict
    return base_validate(path, iteritems(data), out)
  File "/usr/local/lib/python3.8/site-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
    raise er.MultipleInvalid(errors)
voluptuous.error.MultipleInvalid: expected float for dictionary value @ data['timestamp']
```

This may actually be a "frontend" issue, so feel free to move if necessary.

**Comment by probot-home-assistant** [comment]
Hey there @home-assistant/core, mind taking a look at this issue as its been labeled with an integration (`input_datetime`) you are listed as a [codeowner](https://github.com/home-assistant/core/blob/dev/CODEOWNERS#L204) for? Thanks!
<sub><sup>(message by CodeOwnersMention)</sup></sub>

**Comment by rccoleman** [comment]
In fairness, this is probably a frontend or template engine issue and not directly related to the datetime integration.  I'd move it if I could.

**Comment by rccoleman** [comment]
This is still an issue with 0.116, and another user saw the issue templates feeding into the `rgb_color` for a light.  I'd fix or add a more appropriate label if I could because this seems like a general issue with Dev->Services.

Tinkerer pointed out that 0.117 will return native types and may address this and hopefully that will unify the behavior of scripts & the dev tools.

**Comment by balloob** [comment]
The service dev tool does not render templates.

**Comment by rccoleman** [comment]
And yet, it does, just with this quirk :)

![image](https://user-images.githubusercontent.com/860888/96509125-72d0df00-1210-11eb-9fff-950ffdbf7885.png)

![image](https://user-images.githubusercontent.com/860888/96490425-a5211300-11f5-11eb-86ec-2f48602a73de.png)

![image](https://user-images.githubusercontent.com/860888/96490465-b538f280-11f5-11eb-80e4-b6bdf6233a27.png)

**Comment by rccoleman** [comment]
This still doesn't work properly in 0.117, despite the change to have templates return native data types.  I created an `input_number` helper and attempted to set the value using Dev->Services like this:

<img width="630" alt="Screen Shot 2020-10-24 at 1 36 27 PM" src="https://user-images.githubusercontent.com/860888/97093066-fa965f00-15fd-11eb-86ea-c9d62251ba21.png">

The service data is this:

```
entity_id: input_number.test
value: "{{ states('input_number.test')|float + 10.0 }}"
```

And the error is this:

`voluptuous.error.MultipleInvalid: expected float for dictionary value @ data['value']`

It seems like we're in a sort of middle ground where templates *are* evaluated in dev tools, but types aren't being handled properly.  I see people running across this regularly in Discord, and it makes it difficult to support them when the behavior in DevTools doesn't reflect what will happen in a script/automation.  Here's the latest example:

https://discord.com/channels/330944238910963714/672223497736421388/769651662541291530

**Comment by balloob** [comment]
Nah, it's not a middle ground. Dev tools do not support templates. Persistent notification and notify do support templates in their own service handling.

**Comment by bramkragten** [comment]
Should we move this to the frontend as a feature request?

**Comment by balloob** [comment]
Would you expect the frontend to process the data and then call the service? Or would you expect the websocket API call service command to get a flag to "render templates"? 

I kinda like the idea of putting it in the frontend, because we could show the user what the values are that we passed to the service. Transferring.
